File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* less(ListNode* one, ListNode* two) {
        if(one == nullptr) return two;
        else if(two == nullptr) return one;
        else if(one->val <= two->val) return one;
        else return two;
    }
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* temp1 = list1, *temp2 = list2, *temp3 = list1;
        if(list1 == nullptr && list2 == nullptr) return list1;
        if(list1 == nullptr && list2 != nullptr) return list2;
        if(list2 == nullptr && list1 != nullptr) return list1;
        if(list2->val < list1->val) {
            temp3 = list2;
        }
        while(temp1 != nullptr && temp2 != nullptr) {
            if(temp1 == less(temp1,temp2)) {
                temp1 = temp1->next;
                list1->next = less(temp1,temp2);
                list1 = temp1;
            } else {
                temp2 = temp2->next;
                list2->next = less(temp1,temp2);
                list2 = temp2;
            }
        }
        return temp3;
    }
};

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    if (l1 === null && l2 === null) {
        return null;
    } else if (l1 === null) {
        return l2;
    } else if (l2 === null) {
        return l1;
    }
    var dummy;
    var temp;
    if (l1.val > l2.val) {
        dummy = new ListNode(0, l2);
        temp = l2;
        l2 = l2.next;
    } else {
        dummy = new ListNode(0, l1);
        temp = l1;
        l1 = l1.next;
    }
    while (l1 !== null && l2 !== null) {
        if (l1.val > l2.val) {
            temp.next = l2;
            temp = l2;
            l2 = l2.next;
        } else {
            temp.next = l1;
            temp = l1;
            l1 = l1.next;
        }
    }
    if (l1 === null) {
        temp.next = l2;
    }
    if (l2 === null) {
        temp.next = l1;
    }
    return dummy.next;
};